File: ios/Classes/FreshchatPlugin.m
#import "FreshchatPlugin.h"
#import "FreshchatSDK.h"

static const NSString* METHOD_INIT = @"init";
static const NSString* METHOD_IDENTIFY_USER = @"identifyUser";
static const NSString* METHOD_UPDATE_USER_INFO = @"updateUserInfo";
static const NSString* METHOD_RESET_USER = @"reset";
static const NSString* METHOD_SHOW_CONVERSATIONS = @"showConversations";
static const NSString* METHOD_SHOW_FAQS = @"showFAQs";
static const NSString* METHOD_GET_UNREAD_MESSAGE_COUNT = @"getUnreadMsgCount";
static const NSString* METHOD_SETUP_PUSH_NOTIFICATIONS = @"setupPushNotifications";

@implementation FreshchatPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"freshchat"
            binaryMessenger:[registrar messenger]];
  FreshchatPlugin* instance = [[FreshchatPlugin alloc] initWithRegistrar:registrar];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)initWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  self = [super init];
  NSAssert(self, @"super init cannot be nil");
  _registrar = registrar;
  return self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  id viewController = [UIApplication sharedApplication].delegate.window.rootViewController;

  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  }
  else if ([METHOD_INIT isEqualToString:call.method]) {
      NSString* appID = call.arguments[@"appID"];
      NSString* appKey = call.arguments[@"appKey"];
      
      FreshchatConfig* freshchatConfig = [[FreshchatConfig alloc] initWithAppID:appID andAppKey:appKey];
      
      freshchatConfig.cameraCaptureEnabled = false;
      freshchatConfig.responseExpectationVisible = NO; //set NO to hide it if you want to hide the response expectation for the channel

      // NSString* key = [_registrar lookupKeyForAsset:@"FCTheme_New"];
      // NSString* path = [[NSBundle mainBundle] pathForResource:key ofType:nil];
      freshchatConfig.themeName = @"FCTheme_New";
      
      [[Freshchat sharedInstance] initWithConfig:freshchatConfig];
      
      result([NSNumber numberWithBool:YES]);
  }
  else if ([METHOD_SHOW_CONVERSATIONS isEqualToString:call.method]) {
      NSArray* tags = call.arguments[@"tags"];
      NSString* title = call.arguments[@"title"];
      if (tags.count > 0) {
          ConversationOptions* convOptions = [ConversationOptions new];
          [convOptions filterByTags:tags withTitle:title];
          [[Freshchat sharedInstance] showConversations:viewController withOptions:convOptions];
      } else {
          [[Freshchat sharedInstance] showConversations:viewController];
      }
      result([NSNumber numberWithBool:YES]);
  }
  else if ([METHOD_UPDATE_USER_INFO isEqualToString:call.method]) {
      NSString* firstName = call.arguments[@"firstName"];
      NSString* email = call.arguments[@"email"];
      FreshchatUser* freshchatUser = [FreshchatUser sharedInstance];
      
      if (firstName != nil && [firstName length] > 0) {
          freshchatUser.firstName = firstName;
      }
      
      freshchatUser.email = email;
      
      [[Freshchat sharedInstance] setUser:freshchatUser];
      
      result([NSNumber numberWithBool:YES]);
  }
  else if ([METHOD_IDENTIFY_USER isEqualToString:call.method]) {
      NSString* externalId = call.arguments[@"externalID"];
      NSString* restoreId = call.arguments[@"restoreID"];
      
      if ([restoreId isEqualToString:@""]) {
          [[Freshchat sharedInstance] identifyUserWithExternalID:externalId restoreID:nil];
          restoreId = [FreshchatUser sharedInstance].restoreID;
      }
      else {
          [[Freshchat sharedInstance] identifyUserWithExternalID:externalId restoreID:restoreId];
      }
      
      result(restoreId);
  }
  else if ([METHOD_SHOW_FAQS isEqualToString:call.method]) {
      const NSNumber* showFaqCategoriesAsGrid = call.arguments[@"showFaqCategoriesAsGrid"];
      const NSNumber* showContactUsOnAppBar = call.arguments[@"showContactUsOnAppBar"];
      const NSNumber* showContactUsOnFaqScreens = call.arguments[@"showContactUsOnFaqScreens"];
      
      //todo: seems not used on ios.
      //const NSNumber* showContactUsOnFaqNotHelpful = call.arguments[@"showContactUsOnFaqNotHelpful"];
      
      FAQOptions* faqOptions = [FAQOptions new];
      faqOptions.showFaqCategoriesAsGrid = showFaqCategoriesAsGrid;
      faqOptions.showContactUsOnAppBar = showContactUsOnAppBar;
      faqOptions.showContactUsOnFaqScreens = showContactUsOnFaqScreens;
      
      [[Freshchat sharedInstance] showFAQs:viewController withOptions:faqOptions];
      
      result([NSNumber numberWithBool:YES]);
  }
  else if ([METHOD_GET_UNREAD_MESSAGE_COUNT isEqualToString:call.method]) {
      /* If you want to indicate to the user that he has unread messages in his inbox, you can retrieve the unread count to display. */
      //returns an int indicating the of number of unread messages for the user
      [[Freshchat sharedInstance]unreadCountWithCompletion:^(NSInteger count) {
          NSLog(@"your unread count : %d", (int)count);
          result([NSNumber numberWithLong:count]);
      }];
  }
  else if ([METHOD_SETUP_PUSH_NOTIFICATIONS isEqualToString:call.method]) {
      const NSString* token = call.arguments[@"token"];
      [[Freshchat sharedInstance] setPushRegistrationToken:token];
      result([NSNumber numberWithBool:YES]);
  }
  else if ([METHOD_RESET_USER isEqualToString:call.method]) {
      [[Freshchat sharedInstance] resetUserWithCompletion:^{
          result([NSNumber numberWithBool:YES]);
      }];
  }
  else {
    result(FlutterMethodNotImplemented);
  }
}
@end

File: android/src/main/java/eu/smokify/freshchat/FreshchatPlugin.java
package eu.smokify.freshchat;

import android.app.Application;

import com.freshchat.consumer.sdk.ConversationOptions;
import com.freshchat.consumer.sdk.FaqOptions;
import com.freshchat.consumer.sdk.Freshchat;
import com.freshchat.consumer.sdk.FreshchatCallbackStatus;
import com.freshchat.consumer.sdk.FreshchatConfig;
import com.freshchat.consumer.sdk.FreshchatUser;
import com.freshchat.consumer.sdk.UnreadCountCallback;
import com.freshchat.consumer.sdk.exception.MethodNotAllowedException;

import java.util.ArrayList;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FreshchatPlugin */
public class FreshchatPlugin implements MethodCallHandler {
  private static final String METHOD_INIT = "init";
  private static final String METHOD_IDENTIFY_USER = "identifyUser";
  private static final String METHOD_UPDATE_USER_INFO = "updateUserInfo";
  private static final String METHOD_RESET_USER = "reset";
  private static final String METHOD_SHOW_CONVERSATIONS = "showConversations";
  private static final String METHOD_SHOW_FAQS = "showFAQs";
  private static final String METHOD_GET_UNREAD_MESSAGE_COUNT = "getUnreadMsgCount";
  private static final String METHOD_SETUP_PUSH_NOTIFICATIONS = "setupPushNotifications";

  private final Application application;

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "freshchat");
    channel.setMethodCallHandler(new FreshchatPlugin((Application) registrar.context()));
  }

  private FreshchatPlugin(Application application) {
    this.application = application;
  }

  @Override
  public void onMethodCall(MethodCall call, final Result result) {

    switch (call.method) {
      case METHOD_INIT:
        final String appID = call.argument("appID");
        final String appKey = call.argument("appKey");
        FreshchatConfig freshchatConfig = new FreshchatConfig(appID, appKey);
        freshchatConfig.setTeamMemberInfoVisible(false);
        freshchatConfig.setCameraCaptureEnabled(false);
        Freshchat.getInstance(this.application.getApplicationContext()).init(freshchatConfig);
        result.success(true);
        break;
      case METHOD_IDENTIFY_USER:
        final String externalId = call.argument("externalID");
        String restoreId = call.argument("restoreID");

        try {
          if (restoreId == "") {
            Freshchat.getInstance(this.application.getApplicationContext()).identifyUser(externalId, null);
            restoreId = Freshchat.getInstance(this.application.getApplicationContext()).getUser().getRestoreId();
          } else {
            Freshchat.getInstance(this.application.getApplicationContext()).identifyUser(externalId, restoreId);
          }
        } catch (MethodNotAllowedException e) {
          e.printStackTrace();
          result.error("Error while identifying User", "error", e);
        }
        result.success(restoreId);
        break;
      case METHOD_UPDATE_USER_INFO:
        final String firstName = call.argument("firstName");
        final String email = call.argument("email");
        final Map<String, String> customProperties = call.argument("custom_property_list");

        FreshchatUser freshchatUser = Freshchat.getInstance(this.application.getApplicationContext()).getUser();
        if (firstName != null) {
            freshchatUser.setFirstName(firstName);
        }
        freshchatUser.setEmail(email);
        try {
          Freshchat.getInstance(this.application.getApplicationContext()).setUser(freshchatUser);

          if (customProperties != null) {
            Freshchat.getInstance(this.application.getApplicationContext()).setUserProperties(customProperties);
          }
        } catch (MethodNotAllowedException e) {
          e.printStackTrace();
          result.error("Error while setting User", "error", e);
        }
        result.success(true);
        break;
      case METHOD_SHOW_CONVERSATIONS:
        final ArrayList tags = call.argument("tags");
        final String title = call.argument("title");
        if (tags.size() > 0) {
          ConversationOptions convOptions = new ConversationOptions().filterByTags(tags, title);
          Freshchat.showConversations(this.application, convOptions);
        } else {
          Freshchat.showConversations(this.application.getApplicationContext());
        }
        result.success(true);
        break;
      case METHOD_SHOW_FAQS:
        final boolean showFaqCategoriesAsGrid = call.argument("showFaqCategoriesAsGrid");
        final boolean showContactUsOnAppBar = call.argument("showContactUsOnAppBar");
        final boolean showContactUsOnFaqScreens = call.argument("showContactUsOnFaqScreens");
        final boolean showContactUsOnFaqNotHelpful = call.argument("showContactUsOnFaqNotHelpful");

        FaqOptions faqOptions = new FaqOptions().showFaqCategoriesAsGrid(showFaqCategoriesAsGrid)
                .showContactUsOnAppBar(showContactUsOnAppBar).showContactUsOnFaqScreens(showContactUsOnFaqScreens)
                .showContactUsOnFaqNotHelpful(showContactUsOnFaqNotHelpful);

        Freshchat.showFAQs(this.application, faqOptions);
        result.success(true);
        break;
      case METHOD_GET_UNREAD_MESSAGE_COUNT:
        Freshchat.getInstance(this.application.getApplicationContext()).getUnreadCountAsync(new UnreadCountCallback() {
          @Override
          public void onResult(FreshchatCallbackStatus freshchatCallbackStatus, int i) {
            result.success(i);
          }
        });
        break;
      case METHOD_SETUP_PUSH_NOTIFICATIONS:
        final String token = call.argument("token");
        Freshchat.getInstance(this.application.getApplicationContext()).setPushRegistrationToken(token);
        result.success(true);
        break;
      case METHOD_RESET_USER:
        Freshchat.resetUser(this.application.getApplicationContext());
        result.success(true);
        break;
      default:
        result.notImplemented();
    }
  }
}
